lucene/core/src/java/org/apache/lucene/search/FeatureComparatorSource.java

```diff
@@ -0,0 +1,131 @@
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.PostingsEnum;
+import org.apache.lucene.index.Terms;
+import org.apache.lucene.index.TermsEnum;
+import org.apache.lucene.util.BytesRef;
+
+/** Provides a comparator that parses a feature field's values as float and sorts by descending value */
+public class FeatureComparatorSource extends FieldComparatorSource {
+
+  private final BytesRef featureName;
+  private final Float missingValue;
+
+  /**
+   * Creates a {@link FieldComparatorSource} that can be used to sort hits by
+   * the value of a particular feature in a {@link FeatureField}. Note that the
+   * field name to use for the sort will be specified by the {@link SortField}.
+   *
+   * @param featureName The name of the feature to use for the sort value
+   * @param missingValue The value to use for documents that don't have a value
+   *   for the feature. If <code>null</code>, documents with no value for the
+   *   feature will be sorted as if their value is <code>0.0f</code>.
+   */
+  public FeatureComparatorSource(BytesRef featureName, Float missingValue) {
+    this.featureName = featureName;
+    this.missingValue = missingValue;
+  }
+
+  @Override
+  public FieldComparator<?> newComparator(String fieldname, int numHits, int sortPos, boolean reversed) {
+    return new FeatureComparator(numHits, fieldname, featureName, missingValue);
+  }
+
+  /** Parses a feature field's values as float and sorts by descending value */
+  class FeatureComparator extends SimpleFieldComparator<Float> {
+    private final String field;
+    private final BytesRef featureName;
+    private final float missingValue;
+    private final float[] values;
+    private float bottom;
+    private float topValue;
+    private PostingsEnum currentReaderPostingsValues;
+
+    /** Creates a new comparator based on relevance for {@code numHits}. */
+    public FeatureComparator(int numHits, String field, BytesRef featureName, Float missingValue) {
+      this.values = new float[numHits];
+      this.field = field;
+      this.featureName = featureName;
+      this.missingValue = missingValue != null ? missingValue : 0.0f;
+    }
+
+    @Override
+    protected void doSetNextReader(LeafReaderContext context) throws IOException {
+      Terms terms = context.reader().terms(field);
+      if (terms == null) {
+        currentReaderPostingsValues = null;
+      } else {
+        TermsEnum termsEnum = terms.iterator();
+        if (termsEnum.seekExact(featureName) == false) {
+          currentReaderPostingsValues = null;
+        } else {
+          currentReaderPostingsValues = termsEnum.postings(currentReaderPostingsValues, PostingsEnum.FREQS);
+        }
+      }
+    }
+
+    private float getValueForDoc(int doc) throws IOException {
+      if (currentReaderPostingsValues != null
+          && (currentReaderPostingsValues.docID() == doc || currentReaderPostingsValues.advance(doc) == doc)) {
```

Comment:
this might call advance on a target that is less that the current doc (which is illegal) if getValueForDoc is called consecutively on two doc IDs that don't have any match in the postings in between them